File: playlstr/util/export.py
from .playlist import *


def export_as_text(playlist_name: str, tracks: list) -> str:
    text = playlist_name + "\n\n"
    for track in Track.objects.filter(track_id__in=tracks):
        text += track.as_export_text() + "\n"
    return text


def export_as_m3u(playlist_name: str, tracks: list) -> str:
    # TODO
    raise ValueError


def tracks_matching_criteria(playlist_id: int, criteria: dict) -> list:
    try:
        playlist = Playlist.objects.get(playlist_id=playlist_id)
    except ObjectDoesNotExist:
        raise ValueError
    tracks = PlaylistTrack.objects.filter(playlist=playlist).values_list(
        "track", flat=True
    )
    query = Q(track_id__in=tracks)
    if "excluded_genres" in criteria:
        for genre in criteria["excluded_genres"]:
            query &= ~Q(genres__icontains=genre)
    if "excluded_artists" in criteria:
        for artist in criteria["excluded_artists"]:
            query &= ~Q(artist=artist)
    if "excluded_albums" in criteria:
        for album in criteria["excluded_albums"]:
            query &= ~Q(album=album)
    if "explicit" in criteria and not criteria["explicit"]:
        query &= Q(explicit=False)
    return Track.objects.filter(query)

File: playlstr/static/js/export.js
let excludedArtists = [];
let excludedAlbums = [];

function excludeSelectedAlbum() {
    let option = $('#albumExcludeSelect').children('option:selected');
    if (option === $('#albumExcludeSelect:first-child')) return; // Default 'Choose...' element isn't an album
    excludedAlbums.push(option.val());
    option.remove();
    let id = option.val().replace(' ', '_');
    $('#excludedAlbums').append('<div id="{0}" class="form-inline">{1}<button onclick="removeExcludedAlbum(\'{1}\');" type="button" class="btn btn-outline-danger">X</button></div>'.format(id, option.val()));
}

function removeExcludedAlbum(album) {
    let index = excludedAlbums.indexOf(album);
    if (index === -1) return;
    $('#' + album.replace(' ', '_')).remove();
    excludedAlbums.splice(index, 1);
    $('#albumExcludeSelect').append($(new Option(album, album)).html(album));
}

function excludeSelectedArtist() {
    let option = $('#artistExcludeSelect').children('option:selected');
    if (option === $('#artistExcludeSelect:first-child')) return; // Default 'Choose...' element isn't an artist
    excludedArtists.push(option.val());
    option.remove();
    let id = option.val().replace(' ', '_');
    $('#excludedArtists').append('<div id="{0}" class="form-inline">{1}<button onclick="removeExcludedArtist(\'{1}\');" type="button" class="btn btn-outline-danger">X</button></div>'.format(id, option.val()));
}

function removeExcludedArtist(artist) {
    let index = excludedArtists.indexOf(artist);
    if (index === -1) return;
    $('#' + artist.replace(' ', '_')).remove();
    excludedArtists.splice(index, 1);
    $('#artistExcludeSelect').append($(new Option(artist, artist)).html(artist));
}

function exportAsText() {
    let criteria = JSON.stringify({
        'excluded_genres': getExcludedGenres(),
        'explicit': $('#explicitCheck').is(':checked'),
        'excluded_artists': excludedArtists,
        'excluded_albums': excludedAlbums
    });
    $.ajax({
        type: 'POST',
        url: 'http://' + window.location.host + '/file-export/',
        data: {'playlist_id': playlist_id, 'filetype': 'text', 'criteria': criteria},
        headers: {'X-CSRFToken': csrfToken},
        success: function (result) {
            console.log('success');
            console.log(result);
            let blob = new Blob([result], {type: 'text/plain'});
            window.location = URL.createObjectURL(blob);
        },
        error: function () {
            exportFail();
        }
    });
}

function getExcludedGenres() {
    // TODO
    return [];
}

function exportToSpotify() {
    let criteria = JSON.stringify({
        'excluded_genres': getExcludedGenres(),
        'explicit': $('#explicitCheck').is(':checked'),
        'excluded_artists': excludedArtists,
        'excluded_albums': excludedAlbums
    });
    $.ajax({
        type: 'POST',
        url: 'http://' + window.location.host + '/spotify-export/',
        data: {'playlist_id': playlist_id, 'criteria': criteria},
        headers: {'X-CSRFToken': csrfToken},
        success: function (result) {
            window.location = result;
        },
        error: exportFail
    });
}

function exportFail(data) {
    $('#exportError').text('Unable to export playlist (' + data.responseText + ')').fadeIn().delay(4000).fadeOut();
}


